src/tests/unittests/d3d12/CopySplit.cpp

```diff
@@ -0,0 +1,375 @@
+                bool overlapY = RangesOverlap(a.texture.y, a.texture.y + a.texture.height, b.texture.y, b.texture.y + b.texture.height);
+                bool overlapZ = RangesOverlap(a.texture.z, a.texture.z + a.texture.depth, b.texture.z, b.texture.z + b.texture.depth);
+                ASSERT_TRUE(!overlapX || !overlapY || !overlapZ);
+            }
+        }
+    }
+
+    // Check that the union of the copy regions exactly covers the texture region
+    void ValidateTextureCoverage(const TextureSpec& textureSpec, const CopySplit& copySplit) {
+        ASSERT_TRUE(copySplit.count > 0);
+
+        uint32_t minX = copySplit.copies[0].texture.x;
+        uint32_t minY = copySplit.copies[0].texture.y;
+        uint32_t minZ = copySplit.copies[0].texture.z;
+        uint32_t maxX = copySplit.copies[0].texture.x + copySplit.copies[0].texture.width;
+        uint32_t maxY = copySplit.copies[0].texture.y + copySplit.copies[0].texture.height;
+        uint32_t maxZ = copySplit.copies[0].texture.z + copySplit.copies[0].texture.depth;
+
+        for (uint32_t i = 1; i < copySplit.count; ++i) {
+            const CopyInfo::Region& texture = copySplit.copies[i].texture;
+            minX = std::min(minX, texture.x);
+            minY = std::min(minY, texture.y);
+            minZ = std::min(minZ, texture.z);
+            maxX = std::max(maxX, texture.x + texture.width);
+            maxY = std::max(maxY, texture.y + texture.height);
+            maxZ = std::max(maxZ, texture.z + texture.depth);
+        }
+
+        ASSERT_EQ(minX, textureSpec.x);
+        ASSERT_EQ(minY, textureSpec.y);
+        ASSERT_EQ(minZ, textureSpec.z);
+        ASSERT_EQ(maxX, textureSpec.x + textureSpec.width);
+        ASSERT_EQ(maxY, textureSpec.y + textureSpec.height);
+        ASSERT_EQ(maxZ, textureSpec.z + textureSpec.depth);
+    }
+
+    // Check that copies start in the buffer exactly at the buffer offset
+    void ValidateBufferOffset(const TextureSpec& textureSpec, const BufferSpec& bufferSpec, const CopySplit& copySplit) {
+        ASSERT_TRUE(copySplit.count > 0);
+
+        uint32_t offset = (
+            copySplit.copies[0].buffer.x * textureSpec.texelSize +
+            copySplit.copies[0].buffer.y * bufferSpec.rowPitch +
+            copySplit.copies[0].buffer.z * bufferSpec.rowPitch * textureSpec.height);
+
+        for (uint32_t i = 1; i < copySplit.count; ++i) {
+            const CopyInfo::Region& buffer = copySplit.copies[i].buffer;
+            offset = std::min(offset,
+                buffer.x * textureSpec.texelSize +
+                buffer.y * bufferSpec.rowPitch +
+                buffer.z * bufferSpec.rowPitch * textureSpec.height);
+        }
+
+        ASSERT_EQ(copySplit.offset + offset, bufferSpec.offset);
+    }
+   
+    // Validate there are exactly two copies if the copy straddles the row pitch and one copy otherwise
+    void ValidateCopyCount(const TextureSpec& textureSpec, const BufferSpec& bufferSpec, const CopySplit& copySplit) {
```

Comment:
This is very tied to the implementation detail of the split, and the number of copies is kind of an implementation detail (not important for correctness, only for perf). So I think we could remove this check.